packages/jsii-pacmak/lib/targets/dotnet/runtime-type-checking.ts

```diff
@@ -0,0 +1,242 @@
+  }
+}
+
+abstract class Validation {
+  public static forTypeReference(
+    argument: string,
+    expression: string,
+    description: string,
+    ref: TypeReference,
+    allowNull: boolean,
+  ): Validation | undefined {
+    if (ref.unionOfTypes) {
+      return Validation.unionCheck(
+        argument,
+        expression,
+        description,
+        ref.unionOfTypes,
+        allowNull,
+      );
+    } else if (ref.arrayOfType) {
+      return Validation.collectionCheck(
+        argument,
+        expression,
+        description,
+        'array',
+        ref.arrayOfType,
+      );
+    } else if (ref.mapOfType) {
+      return Validation.collectionCheck(
+        argument,
+        expression,
+        description,
+        'map',
+        ref.mapOfType,
+      );
+    }
+    return undefined;
+  }
+
+  public static collectionCheck(
+    argument: string,
+    expression: string,
+    description: string,
+    type: 'array' | 'map',
+    elementType: TypeReference,
+  ): Validation | undefined {
+    const elementValidator = Validation.forTypeReference(
+      argument,
+      `${expression}[idx]`,
+      `${description}[@{idx}]`,
+      elementType,
+      false,
+    );
+    if (elementValidator == null) {
+      return undefined;
+    }
+
+    class CollectionCheck extends Validation {
+      public emit(code: CodeMaker, resolver: DotNetTypeResolver): void {
+        // We need to come up with a unique-enough ID here... so we use a hash.
+        const prefix = type === 'array' ? '__idx' : '__item';
+        const varName = `${prefix}_${createHash('sha256')
+          .update(expression)
+          .digest('hex')
+          .slice(0, 6)}`;
+        if (type === 'array') {
+          code.openBlock(
+            `for (var ${varName} = 0 ; ${varName} < ${expression}.Length ; ${varName}++)`,
+          );
+        } else {
+          code.openBlock(`foreach (var ${varName} in ${expression})`);
+        }
+        Validation.forTypeReference(
+          argument,
+          type === 'array' ? `${expression}[${varName}]` : `${varName}.Value`,
+          `${description}[${
+            type === 'array' ? `{${varName}}` : `"{${varName}.Key}"`
+          }]`,
+          elementType,
+          false,
+        )!.emit(code, resolver);
+        code.closeBlock();
+      }
+    }
+
+    return new CollectionCheck();
+  }
+
+  public static unionCheck(
+    argument: string,
+    expression: string,
+    description: string,
+    types: readonly TypeReference[],
+    allowNull: boolean,
+  ): Validation {
+    const hasInterface = types.some((t) => t.type?.isInterfaceType());
+
+    class UnionCheck extends Validation {
+      public emit(code: CodeMaker, resolver: DotNetTypeResolver): void {
+        const validTypes = new Array<string>();
+
+        const castVarName = `cast_${createHash('sha256')
+          .update(expression)
+          .digest('hex')
+          .slice(0, 6)}`;
+
+        code.openBlock(`switch (${expression})`);
```

Comment:
Do we want this to be consistent across languages when possible? Should I port this `switch` approach to Java as well, or leave it as-is?

Reply:
You can't do `switch` like this in Java... it only works with primitive types, enums, String, and certain boxed primitive classes. So you end up having to stick to `if`/`else` here, sorry!